bindings/cs/MongoDB.Crypt/CryptClient.cs

```diff
@@ -0,0 +1,252 @@
+                    }
+                }
+            }
+
+            handle.Check(_status, Library.mongocrypt_ctx_setopt_algorithm(handle, Helpers.AlgorithmToString(algorithm), -1));
+
+            unsafe
+            {
+                fixed (byte* p = command)
+                {
+                    IntPtr ptr = (IntPtr)p;
+                    using (PinnedBinary pinned = new PinnedBinary(ptr, (uint)command.Length))
+                    {
+                        handle.Check(_status, Library.mongocrypt_ctx_explicit_encrypt_init(handle, pinned.Handle));
+                    }
+                }
+            }
+
+            return new CryptContext(handle);
+        }
+
+        /// <summary>
+        /// Starts the decryption context.
+        /// </summary>
+        /// <param name="buffer">The bson document to decrypt.</param>
+        /// <returns>A decryption context</returns>
+        public CryptContext StartDecryptionContext(byte[] buffer)
+        {
+            ContextSafeHandle handle = Library.mongocrypt_ctx_new(_handle);
+
+            GCHandle gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);
+
+            unsafe
+            {
+                fixed (byte* p = buffer)
+                {
+                    IntPtr ptr = (IntPtr)p;
+                    using (PinnedBinary pinned = new PinnedBinary(ptr, (uint)buffer.Length))
+                    {
+                        handle.Check(_status, Library.mongocrypt_ctx_decrypt_init(handle, pinned.Handle));
+                    }
+                }
+            }
+
+            return new CryptContext(handle);
+        }
+
+        /// <summary>
+        /// Starts an explicit decryption context.
+        /// </summary>
+        /// <param name="buffer">The buffer.</param>
+        /// <returns>A encryption context</returns>
+        public CryptContext StartExplicitDecryptionContext(byte[] buffer)
+        {
+            ContextSafeHandle handle = Library.mongocrypt_ctx_new(_handle);
+
+            unsafe
+            {
+                fixed (byte* p = buffer)
+                {
+                    IntPtr ptr = (IntPtr)p;
+                    using (PinnedBinary pinned = new PinnedBinary(ptr, (uint)buffer.Length))
+                    {
+                        // Let mongocrypt run strlen
+                        handle.Check(_status, Library.mongocrypt_ctx_explicit_decrypt_init(handle, pinned.Handle));
+                    }
+                }
+            }
+
+            return new CryptContext(handle);
+        }
+
+
+        void IStatus.Check(Status status)
+        {
+            Library.mongocrypt_status(_handle, status.Handle);
+        }
+
+        #region IDisposable
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_handle.IsClosed)
```

Comment:
Why only if IsClosed?

Almost seems like it should be if NOT IsClosed?

Reply:
Fixed!